Modified the retrieval of fields to adjust for html

dorrin-sot opened the pull request:
## Purpose / Description
Fixed bug where if the fields in a note contained html characters `<,>,",',￠,£,¥,€,©,®`, it wouldn't display them properly. Basically counting them as part of the actual html tags used to display contents of a note

## Fixes
Fixes #8784 

## Approach
Added method `adjustStringForHtml(String str)` in class `Card` that would replace said characters with their proper html equivalents except for actual tags. Made `getFields` return  a stream of mFields mapped to `adjustStringForHtml(field)`.

## How Has This Been Tested?
Testing out said characters in different types of cards and checking what is displayed.

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

prateek-singh-3212 reviewed (changes requested):
@dorrin-sot  Your FIX looks good to me. Can you share the screenshot of the FIX so that it will be easier for us to review? and secondly, your commit history is very confusing because of the same commit message. So please rebase the commits into a single commit

dorrin-sot commented:
@prateek-singh-3212

https://user-images.githubusercontent.com/59933477/118866789-69693b80-b8f3-11eb-9834-c340e0bb5721.mov

dorrin-sot commented:
About the commits, I just went back and forth between using different ways to pass the tests and later realized it was unnecessary but in the end all I changed was:
made "getFields" return a list of fields that were adjust for the html format of displaying the cards. so for example if someone had "<" in their note but not "<tag/>" it would replace it with "&lt;".
plus I changed one of the tests that was about multiline cloze tests that was already using "contains" abit so it wouldn't fail because of the "getFields" change

mikehardy commented:
Similar to David's concern on #8870 (which duplicates this) - I would like to make sure we are doing the same thing as upstream. The problem is I can't *find* this upstream code. @dae - can I ping you for help? I would love to see where the tags are html-entity-ified for cloze but I'm not sure where that is and I attempted to look in pylib and rslib on ankitects :thinking:

dae commented:
The desktop uses a WYSIWYG editor, so if a user types in a < it gets converted to `&lt;` automatically by the web toolkit. If my understanding of this PR is correct, it seems to be less about fixing a "bug" so much as attempting to fix invalid HTML input by the user, and trying to do that with string replacements seems like it could have unintended side-effects

mikehardy commented:
Hmm.

So there is a sort of root cause where people may type busted HTML entities into a field, which AnkiDesktop presumes won't happen, but AnkiDesktop may have to deal with busted input from non-WYSIWYG clients (like AnkiDroid), will it have the same issue? I don't think we can ever assume the contents of a cloze will have input with all HTML entities escaped

Is it safe to say that the contents of a cloze will not contain actual HTML? I don't think it is, I think you'll allow styling and bolding and such correct?

So we could have a mix of real HTML and bogus HTML. I'll wait to hear if my assumption HTML is valid inside a close is correct before thinking on this more

dorrin-sot commented:
@dae @mikehardy I just realized my code doesn't work for cards that have a mix of html and non-html text like `a<b&yes a is less than b <b>bold<b/><em>italic<em/>`. As @dae said, I see you can't write html code directly into the fields in Anki Desktop as It would all get escaped. If it's ok, I can do that for Ankidroid as well by adding a button for editing the html code directly🤔 or I can also just make it so that html code has to be surrounded by a certain delimiter like `` in github.

Plus, I have to add that this issue was not limited to only Cloze type notes

dae commented:
Re your question Mike, cloze deletions can contain HTML inside them (users bolding content after it is clozed for example).

mikehardy commented:
Okay, so if AnkIDesktop has a WYSIWYG editor and will potentially add valid HTML inside cloze (or any other field in a note) *and* will escape non-formatting HTML entities, then we have to assume AnkiDesktop won't generate cards like this by default. But obviously typing them in manually on AnkiDroid makes them likely.

A related part of my thoughts above is how AnkiDesktop handles cards with non-escaped HTML entities that result in invalid HTML. @dorrin-sot in your video above you show that AnkiDesktop is apparently handling the cards okay - how is it doing that?

dorrin-sot commented:
>  @dorrin-sot in your video above you show that AnkiDesktop is apparently handling the cards okay - how is it doing that?

I'm not sure about the code used by AnkiDesktop though I imagine It escapes every character entered at the moment it is typed and when trying to do something like bolding it adds `<b></b>` directly in the html code though it can't be seen in the editor. Then If the user wishes to edit the html code they tap the button to do so (Like below).


https://user-images.githubusercontent.com/59933477/119517409-a36d8e00-bd88-11eb-95d0-3c323b817507.mov

dorrin-sot commented:
found [this repo](https://reposhub.com/android/textview-edittext/wasabeef-richeditor-android.html) while looking around for wysiwyg editors for android. is it ok if I replace the FieldEditTexts with these (or maybe extend them)?

mikehardy commented:
Hmm - that does look like an interesting editor, just going off github activity it seems "alive" enough to me that it is not immediately disqualified, and it has a lot of features - @david-allison-1 what do you think about the github rich text editor linked above :arrow_up:

david-allison-1 commented:
We won't know where the strengths and weaknesses are with each editor until we implement them. It's typically going to be a "pick the best of a questionable bunch"

I'm frustrated at myself that I haven't made progress on @Akshay0701 's PR. 

I don't know on this one, but I feel the plan of action should be:

* Get something in
  * Akshay's PR
* List and iterate on the failures (if any)

This may end up meaning that we select another library at a later date, and we should ensure that the libraries that we use allow this to occur.
